From 426eeac556f49c03fd114080016a641dc8867e0f Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 11 Jun 2021 09:49:37 +0300
Subject: [PATCH] [SW-33038] greco: erratum: block access to bootrom

Access to boot rom and fw sram  is to be blocked as a workaround for
H5 erratum GOYA2_0303.

This patch implements the range register for the same.

Change-Id: Id19ffd4b3d2652c1efec6730b9d36aafb7bca7d9
Signed-off-by: Karthikeyan K <karthikk@habana.ai>
---

diff --git a/arch/greco/common.h b/arch/greco/common.h
index 25c6926..4ed4d88 100644
--- a/arch/greco/common.h
+++ b/arch/greco/common.h
@@ -33,6 +33,9 @@
 #define HL_CFG_BASE			0x7FFC000000ull
 #define SDRAM_BASE			0x7FF0000000ull
 #define DRAM_BASE			0x1000000000000ull
+#define BOOT_ROM_BASE			0x7FFB040000ull
+#define BOOT_ROM_SIZE			0x10000
+#define FW_SRAM_SIZE			0x10000
 
 #define HOST_PHYS_BASE			0x2000000000000ull
 #define HOST_PHYS_SIZE			0x2000000000000ull
diff --git a/arch/greco/security.c b/arch/greco/security.c
index 9ee0be3..80258d8 100644
--- a/arch/greco/security.c
+++ b/arch/greco/security.c
@@ -70,6 +70,7 @@
 };
 
 enum greco_sec_if_hbw_rrs {
+	HBW_RR_BOOTROM,
 	HBW_MAX_RR_IDX = 10,
 };
 
@@ -351,7 +352,7 @@
 		preboot_secure_hbw_rr_full_if(rr, min, max, (enum preboot_sec_engine_type)i);
 }
 
-__unused static void preboot_secure_hbw_rr_short(int rr, u64 min64, u64 max64)
+static void preboot_secure_hbw_rr_short(int rr, u64 min64, u64 max64)
 {
 	u32 min = (min64 & ~SDRAM_BASE) >> SHIFT_4K_POS;
 	u32 max = (max64 & ~SDRAM_BASE) >> SHIFT_4K_POS;
@@ -390,6 +391,7 @@
 
 static void preboot_secure_hbw(void)
 {
+	u64 min, max;
 	/*
 	 * There are 10(0..9 rr idx) RRs for HBW
 	 * NOTE: All the RRs for HBW are of 4KB granularity
@@ -405,6 +407,14 @@
 	 *
 	 * rr idx type : greco_sec_if_hbw_rrs
 	 */
+
+	/*
+	 * Workaround for H5 Errata: GOYA2_0303
+	 * Block Boot ROM and FW SRAM : RR(HBW_RR_BOOTROM)
+	 */
+	min = BOOT_ROM_BASE;
+	max = BOOT_ROM_BASE + BOOT_ROM_SIZE + FW_SRAM_SIZE;
+	preboot_secure_hbw_rr_short(HBW_RR_BOOTROM, min, max);
 }
 
 void preboot_security_init(void)
